src/loaders/TiffLoader.js

```diff
@@ -0,0 +1,252 @@
+import { isInTileBounds } from './utils';
+
+/**
+ * This class serves as a wrapper for reading tiff data.
+ * @param {Object} args
+ * @param {Object} args.tiff geotiffjs tiff object.
+ * @param {Object} args.pool Pool that implements a `decode` function.
+ * @param {Object} args.firstImage First image (geotiff Image object) in the tiff (containing base-resolution data).
+ * @param {Object} args.dimensions Dimensions for the loader.
+ * @param {Object} args.metadata The metadata object
+ * @param {Array} args.offsets The offsets of each IFD in the tiff.
+ * @param {Boolean} args.isRgb Whether or not this tiff represents an rgb image.
+ * @param {Boolean} args.isInterleaved Whether or not this tiff represents an interleaved image.
+ * @param {String} args.dtype The dtype of the image like <u2 or <f4.
+ * @param {Object} args.physicalSizes An object like { x: { value, unit }, y: { value, unit } } where the first keys,
+ * x and y for example, are from the `dimensions and the value is the length and the unit is the units for the value
+ * to be interpreted in, like mm for millimeters.
+ * */
+export default class TiffLoader {
+  constructor({
+    tiff,
+    pool,
+    firstImage,
+    dimensions,
+    offsets,
+    metadata,
+    isRgb,
+    isInterleaved,
+    dtype,
+    physicalSizes
+  }) {
+    this.physicalSizes = physicalSizes;
+    // get first image's description, which contains OMEXML
+    this.metadata = metadata;
+    this.dimensions = dimensions;
+    this.dtype = dtype;
+    this.isRgb = isRgb;
+    this.isInterleaved = isInterleaved;
+    this.photometricInterpretation =
+      firstImage.fileDirectory.PhotometricInterpretation;
+    this.tileSize = firstImage.getTileWidth();
+    const { SubIFDs } = firstImage.fileDirectory;
+    this.numLevels = SubIFDs?.length
+      ? SubIFDs.length + 1
+      : this.metadata.getNumberOfImages();
+    this.isPyramid = this.numLevels > 1;
+    this._offsets = offsets || [];
+    this._pool = pool;
+    this._tiff = tiff;
+  }
+
+  /**
+   * Returns a tiff image object for a given loader selection + pyramid level.
+   * This is implemented by all classes that inherit from the tiff loader.
+   * @param {Object} loaderSelection A lodaer selection
+   * @param {number} z Pyramidal resolution level (default is implemented by loader, should be 0 unless noted otherwise).
+   * @returns {Object} Tiff Image object containing parsed IFD.
+   */
+  // eslint-disable-next-line class-methods-use-this,no-unused-vars,no-empty-function
+  async getImages(loaderSelection, z) {}
```

Comment:
Default implementation should throw an error. This just returns a promise for now that resolves to undefined which might be hard to debug.